Default buffer type to 'heap' for 9.0 (#16500)
elastic/logstash: docs/static/config-details.asciidoc

Keep these memory requirements in mind as you calculate your ideal memory allocation.

[[off-heap-buffers-allocation]]
===== Buffer Allocation types
Input plugins such as {agent}, {beats}, TCP, and HTTP will allocate buffers in Java heap memory to read events from the network.

> This is the preferred allocation method as it facilitates debugging memory usage problems (such as leaks and Out of Memory errors) through the analysis of heap dumps.

Reviewer: ```suggestion
Heap memory is the preferred allocation method, as it facilitates debugging memory usage problems (such as leaks and Out of Memory errors) through the analysis of heap dumps.
```

Reviewer: For clarity and better SEO performance

> Before version 9.0.0, Logstash defaulted to allocate direct memory for this purpose instead of heap. To re-enable the previous behaviour {ls} provides

Reviewer: ```suggestion
Before version 9.0.0, {ls} defaulted to direct memory instead of heap for this purpose. To re-enable the previous behavior {ls} provides
```

Reviewer: I rearrange the words with the intention of simplifying the sentence. Please make sure that I didn't change the meaning.  
Also, Elastic standard is US spelling.

a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate
memory buffers for plugins that use them.

> There shouldn't be a noticeable performance impact in switching between `direct` and `heap`. 

Reviewer: ```suggestion
Performance should not be noticeably affected if you switch between `direct` and `heap`. 
```

> While copying bytes from OS buffers to direct memory buffers is faster, Logstash Event objects produced by these plugins end up 
> being allocated on the Java Heap incurring the cost of copying from direct memory to heap memory, regardless of the setting.

Reviewer: ```suggestion
While copying bytes from OS buffers to direct memory buffers is faster, {ls} Event objects produced by these plugins are allocated on the Java Heap, incurring the cost of copying from direct memory to heap memory, regardless of the setting.
```

Keep these memory requirements in mind as you calculate your ideal memory allocation.

[[off-heap-buffers-allocation]]
===== Buffer Allocation types

> Plugins such as {agent}, {beats}, TCP, and HTTP inputs, use by default Java heap memory to allocate the buffers
> used to read events from the network. This is the preferred allocation type because permit to use better tools
> to investigate potential issues with leaking of memory that would result in Out of Memory (OOM) errors.

Reviewer 2: in Asciidoc, there is a convention to have line-breaks between sentences, which makes it a bit easier to trace history in line-by-line diffs:

e.g., 

```suggestion
Plugins such as {agent}, {beats}, TCP, and HTTP inputs, use by default Java heap memory to allocate the buffers used to read events from the network.
This is the preferred allocation type because permit to use better tools to investigate potential issues with leaking of memory that would result in Out of Memory (OOM) errors.
```

Separately, this bit might use some re-arranging for native english grammar:

```suggestion
Plugins such as {agent}, {beats}, TCP, and HTTP inputs use Java heap memory by default to allocate the buffers that they use to read events from the network.
This is the preferred allocation type because it enables us to use better tools
to investigate potential issues with leaking of memory that would result in Out of Memory (OOM) errors.
```

> ===== Buffer Allocation types
> Plugins such as {agent}, {beats}, TCP, and HTTP inputs, use by default Java heap memory to allocate the buffers
> used to read events from the network. This is the preferred allocation type because permit to use better tools
> to investigate potential issues with leaking of memory that would result in Out of Memory (OOM) errors.

Reviewer 3: ```suggestion
===== Buffer Allocation types
Input plugins such as {agent}, {beats}, TCP, and HTTP will allocate buffers in Java heap memory to read events from the network.
This is the preferred allocation method as it facilitates debugging memory usage problems (such as leaks and Out of Memory errors) through the analysis of heap dumps.
```

> However, in the past the default allocation was in direct memory, to re-enable the old behaviour {ls} provides
> a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate
> memory buffers for plugins that use them.

Reviewer 3: ```suggestion
Before version 9.0.0, Logstash defaulted to allocate direct memory for this purpose, instead of heap. To re-enable the previous behaviour {ls} provides
a `pipeline.buffer.type` setting in <<logstash-settings-file>>.
```

> However, in the past the default allocation was in direct memory, to re-enable the old behaviour {ls} provides
> a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate
> memory buffers for plugins that use them.
> Currently it is set to `heap` by default, but you can change it to `direct` to use direct memory space instead.

Reviewer 3: ```suggestion
However, in the past the default allocation was in direct memory, to re-enable the old behaviour {ls} provides
a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate
memory buffers for plugins that use them.
Currently it is set to `heap` by default, but you can change it to `direct` to use direct memory space instead.
```

Author: This is redundant with https://github.com/elastic/logstash/pull/16500#discussion_r1803364600

> Currently it is set to `heap` by default, but you can change it to `direct` to use direct memory space instead.

Reviewer 3: This sentence is redundant 
```suggestion
```

> Performance-wise there shouldn't be a noticeable impact in switching to `direct`, since while direct memory IO is faster,
> Logstash Event objects produced by these plugins end up being allocated on the Java Heap,
> incurring the cost of copying from direct memory to heap memory regardless of the setting.

Reviewer 2: Don't we use pooled buffers anyway? I'm under the impression that _allocations_ of buffers may be faster in direct, but that our pool usage is a primary driver for our buffer use being performant. We also know that large buffer allocations can succeed in heap where they would fail in direct because the GC will rearrange the existing object space to ensure that it _can_ allocate while the direct allocation will simply not have an appropriately-sized continuous chunk. I would rather a small performance cost in this case than a crashing OOM :)

```suggestion
Performance-wise there shouldn't be a noticeable impact in switching to `direct`.
While allocating direct memory for an individual buffer is faster, these plugins use buffer pools to reduce allocations, and heap-managed allocations are significantly safer.
Logstash Event objects produced by these plugins also end up being allocated on the Java Heap, incurring the cost of copying from buffers to heap memory regardless of the setting.
```

Author: We always use the Netty PooledAllocator, but instead of using direct memory buffers it uses Java heap byte[]

Author: Hi @yaauie 
> While allocating direct memory for an individual buffer is faster,

The is not the allocation that's faster, it's the transfer from OS buffers to direct buffers that's faster, that's usually needed when data is moved around in OS, so from network to filesystem, or network to network. In our case that direct buffers content always flow into Java heap space to inflate Logstash's events, so we loose the benefit of direct buffer, or at least is not so dominant.

Under heavy load, namely large number of connections and large messages, the direct memory space can be exhausted and lead to Out of Memory (OOM) errors in off-heap space.

An off-heap OOM is difficult to debug, so {ls} provides a `pipeline.buffer.type` setting in <<logstash-settings-file>> that lets you control where to allocate memory buffers for plugins that use them. 

> Currently it is set to `heap` by default, but you can change it to `direct` to use direct memory space instead.

Reviewer 3: ```
It is important to note that the Java heap sizing requirements will be impacted by this change since
allocations that previously resided on the direct memory will use heap instead. 
Performance-wise there shouldn't be a noticeable impact, since while direct memory IO is faster, Logstash Event objects produced by these plugins end up being allocated on the Java Heap, incurring the cost of copying from direct memory to heap memory regardless of the setting.
```

These must be changed as well, same with:

```
* When you set `pipeline.buffer.type` to `heap`, consider incrementing the Java heap by the 
amount of memory that had been reserved for direct space. 
```
